torch_glow/src/PyTorchFileLoader.cpp

```diff
@@ -15,30 +15,182 @@
+
+  void set(glow::Function *f, std::vector<glow::Placeholder *> *in,
+           std::vector<glow::Placeholder *> *out) {
+    function = f;
+    inputPlaceholders = in;
+    outputPlaceholders = out;
+  }
+
+  void reset() { set(nullptr, nullptr, nullptr); }
+};
+
+/// Meyer's singleton for static fusion symbol.
+static at::Symbol getFusionSymbol() {
+  /// Fusion pass unique symbol.
+  static const auto fusionSymbol =
+      at::Symbol::fromQualString("glow::LoaderFusionPass");
+  return fusionSymbol;
+}
+
+/// Local thread variable, used in PyTorch Custom Fusion pass.
+static thread_local LocalFusionFunction localFusionInfo;
+
+/// Loads JIT Graph into Glow Function.
+llvm::Error loadJitGraphToGlowFunction(
+    torch::jit::Stack &stack, torch::jit::Graph &graph, glow::Function &f,
+    std::vector<glow::Placeholder *> &inputPlaceholders,
+    std::vector<glow::Placeholder *> &outputPlaceholders) {
+  const auto &graphInputs = graph.inputs();
+  const auto numInputs = graphInputs.size();
+  auto inputs = torch::jit::last(stack, numInputs);
+
+  // Load JIT Graph into Glow Function.
+  RETURN_IF_ERR(PyTorchModelLoader::loadJITGraph(
+      f, graph, inputs, inputPlaceholders, outputPlaceholders));
+
+  // Remove from stack input parameters.
+  torch::jit::drop(stack, numInputs);
+  // Lookup placeholders for the output shapes.
+  for (auto *ph : outputPlaceholders) {
+    std::vector<int64_t> sizes;
+    for (auto size : ph->dims()) {
+      sizes.push_back(static_cast<int64_t>(size));
+    }
+    // Create an empty tensor with the correct shape.
+    auto ptT = at::empty(sizes);
+    auto var = torch::autograd::make_variable(ptT);
+    // Add to stack output parameter.
+    stack.push_back(at::IValue(var));
+  }
+
+  return llvm::Error::success();
+}
+
+/// Runs Module forward pass, triggers custom fusion pass if local thread
+/// Glow function is set.
+llvm::Error
+evaluateModuleGraph(std::shared_ptr<torch::jit::script::Module> &module,
+                    std::vector<torch::jit::IValue> &inputs) {
+  try {
+    module->forward(inputs);
+  } catch (const std::exception &x) {
+    RETURN_ERR(x.what());
+  }
+  return llvm::Error::success();
+}
+
+/// Helper struct, which on constructor registers custom fusion pass
+/// and operator.
+struct RegisterCustomFusionPass {
+  RegisterCustomFusionPass() {
+    auto options = c10::OperatorOptions();
+    options.setAliasAnalysis(at::AliasAnalysisKind::PURE);
+    torch::jit::RegisterOperators op({torch::jit::Operator(
+        getFusionSymbol(),
+        [](const torch::jit::Node *node) {
+          return [node](torch::jit::Stack &stack) {
+            // Get JIT Graph.
+            auto graph = node->g(at::attr::Subgraph);
+            auto err = loadJitGraphToGlowFunction(
+                stack, *graph, *localFusionInfo.function,
+                *localFusionInfo.inputPlaceholders,
+                *localFusionInfo.outputPlaceholders);
+            if (static_cast<bool>(err)) {
```

Comment:
glow errToBool

Reply:
I don't want to output error, instead just to throw exception.

Reply:
I see ok